### PR TITLE
NUI-919 retry for longer on multipart uploads

### PR DESCRIPTION
**Pull request by jdelbick**
## Description
We were seeing a lot of errors without retrying for `worker-transfer` uploads. It was using the default value of 60s for `maxRetryDuration` so if the first upload request took longer than 1 min (which if often did during a large load). The easiest solution was to increase maxRetryDuration to 15 mins.

### Background
It is very clear from this unit test & other logs in `worker-transfer` that we do retry on connection errors. 
[example logs](https://adobeiort.splunkcloud.com/en-US/app/search/search?q=search%20index%3D%22runtime_prod_ext%22%20activation_id%3D4e8f47863a2749dd8f47863a27b9dde7%20%7C%20sort%20_time%20%7C%20fields%20%2B%20_time%2Clog_message%20%7C%20fields%20-%20_raw%20%7C%20outputtext&display.page.search.mode=smart&dispatch.sample_ratio=1&earliest=-1d%40d&latest=%40d&display.page.search.showFields=0&display.events.type=table&display.page.search.tab=statistics&display.general.type=statistics&sid=1605217790.273614)

```

2020-11-11 17:27:46.640 | [rendition 0] uploading rendition /nodejsAction/SFWmt9HT/work/4e8f47863a2749dd8f47863a27b9dde7/out/rendition0.transfer as multi-part to https://sa0204030shared0a8682e17.blob.core.windows.net and 21 more urls, size = 221582232 |  
-- | -- | --
2020-11-11 17:27:47.645 | Waiting 219 milliseconds to attempt retry 1, failure: PUT 'https://sa0204030shared0a8682e17.blob.core.windows.net/aem-blob-ns-team-aem-cm-prd-n16403-cm-p20403-e54705/eca6-2310-61c8-4744-802c-9552dc9a0ff5-1605144460590?sig=xez8yDg%2Bdyb5Rx%2FiqmACQ51zVr69yYU4YbIZO4IcWqM%3D&se=2020-11-12T13%3A27%3A40Z&sv=2019-02-02&sp=w&sr=b&blockId=MDAwMDAz&comp=block' connect failed: request to https://sa0204030shared0a8682e17.blob.core.windows.net/aem-blob-ns-team-aem-cm-prd-n16403-cm-p20403-e54705/eca6-2310-61c8-4744-802c-9552dc9a0ff5-1605144460590?sig=xez8yDg%2Bdyb5Rx%2FiqmACQ51zVr69yYU4YbIZO4IcWqM%3D&se=2020-11-12T13%3A27%3A40Z&sv=2019-02-02&sp=w&sr=b&blockId=MDAwMDAz&comp=block failed, reason: read ECONNRESET |  
2020-11-11 17:27:48.415 | Attempting retry 1 after waiting 219 milliseconds.
```

However, we use the default retry settings:
```
maxRetryDuration = 60000 ms (1min)
```
We sometimes see connection errors from the first upload request that lasts >60s
[metrics](https://insights.newrelic.com/apps/accounts/2045220/asset-compute-production?query=SELECT%20average(duration),%20percentile(duration,%2095),%20percentile(duration,%2099)%20from%20http%20where%20actionName%3D%27worker-transfer%27%20and%20package%3D%27it-20201112-nui-worker-transfer-50c7a77aa5ef%27%20and%20method%3D%27PUT%27%20since%2090%20days%20ago)
Screen Shot 2020-11-12 at 4 43 13 PM

We can increase the `maxRetryDuration` for multipart upload. Since it's for upload, I would suggest “best effort”, meaning we try until the action times out…however some action timeouts are 1 hour so this could waste a lot of resources and we don't have any issues right now with short timeouts besides in `worker-transfer`

### Other Changes
- add warning log when memory usage passes 90% (this can be moved to a separate PR if requested)
[example logs of it working](https://adobeiort.splunkcloud.com/en-US/app/search/search?q=search%20index%3D%22runtime_prod_ext%22%20%22WARNING%3A%20Memory%22&display.page.search.mode=smart&dispatch.sample_ratio=1&earliest=%40d&latest=now&display.page.search.showFields=0&display.events.type=table&display.page.search.tab=statistics&display.general.type=events&sid=1605299065.279001)

Fixes # https://jira.corp.adobe.com/browse/NUI-919

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] I have signed the [Adobe Open Source CLA](https://opensource.adobe.com/cla.html).
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: lib/storage/http.js [hunk 1]
```diff
@@ -16,6 +16,8 @@ const http = require('@adobe/httptransfer');
 const { AssetComputeLogUtils, GenericError, RenditionTooLarge } = require('@adobe/asset-compute-commons');
 const { actionName } = require('../action');
 
+const MAX_RETRY_DURATION_UPLOAD = 900000; // 15 mins
```
Comment (jdelbick) [inline lib/storage/http.js:19]: note: before this change, this value was the default 60s

yes we could get fancy and determine this value based on how much time is left in the action, etc, but I went with the simplest solution that fixes the targetted bug

File: lib/storage/http.js [hunk 2]
```diff
@@ -59,6 +61,7 @@ async function upload(rendition) {
             console.log(`uploading rendition ${file} as multi-part to ${AssetComputeLogUtils.redactUrl(target.urls[0])} and ${target.urls.length-1} more urls, size = ${rendition.size()}`);
             await http.uploadAEMMultipartFile(file, target, {
                 retryEnabled: !process.env.ASSET_COMPUTE_DISABLE_RETRIES,
+                retryMaxDuration: process.env.ASSET_COMPUTE_TEST_RETRY_DURATION || MAX_RETRY_DURATION_UPLOAD,
```
Comment (jdelbick) [inline lib/storage/http.js:64]: note: `process.env.ASSET_COMPUTE_TEST_RETRY_DURATION` is used for unit tests

File: lib/storage/http.js (hunk 1 above)
Comment (tmathern) [reply lib/storage/http.js:19]: ACK

File: lib/worker.js [hunk 3]
```diff
@@ -132,6 +132,15 @@ class AssetComputeWorker {
         if (!process.env.ASSET_COMPUTE_SDK_DISABLE_CGROUP_METRICS) {
             this.cgroupSampler = new Sampler(() => {
                 const metrics_object = cgroupMetrics();
+                const memoryUsage = metrics_object.memory.containerUsagePercentage;
```
Comment (tmathern) [inline lib/worker.js:135]: I think we should remove that before merging to master.
This was more a trial hack - we should come up with something cleaner and nicer later.

**Review by tmathern — APPROVED** [review]
Approving - but remove the high-memory usage logging.

**Pull request merged**